{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#add all tested fuctions\n",
    "\n",
    "def matching(index1, index2):\n",
    "    matched = False\n",
    "    inverse = \"\"\n",
    "    \n",
    "    for x in index2:\n",
    "        if x == \"A\":\n",
    "            inverse =  \"T\" + inverse \n",
    "        elif x == \"T\":\n",
    "            inverse =  \"A\" + inverse\n",
    "        elif x == \"G\":\n",
    "            inverse = \"C\" + inverse\n",
    "        elif x == \"C\":\n",
    "            inverse = \"G\" + inverse   \n",
    "        else:\n",
    "            inverse = \"N\" + inverse\n",
    "\n",
    "#assuming N base calls are normally on the first base\n",
    "#first 100 reads in both files only have Ns on first base\n",
    "#this allows one base pair difference between matching indexes\n",
    "#keeps down the rate of false positives \n",
    "\n",
    "    if containsN(index1):\n",
    "        index1 = index1[1:]\n",
    "        inverse = inverse[1:]\n",
    "        \n",
    "    elif containsN(index2):\n",
    "        index1 = index1[:-1]\n",
    "        inverse = inverse[:-1]\n",
    "        \n",
    "    elif containsN(index1) and containsN(index2):\n",
    "        index1 = index1[1:-1]\n",
    "        inverse = inverse[1:-1]\n",
    "        \n",
    "    return (inverse == index1)\n",
    "\n",
    "def writeToBadFile(outputFile, read, reason):\n",
    "    N = 1\n",
    "    IH = 2\n",
    "    QS = 4\n",
    "    \n",
    "    if (reason & N):\n",
    "        read[2] = read[2] + \" \" + \"(N)\"\n",
    "        \n",
    "    if (reason & IH):\n",
    "        read[2] = read[2] + \" \" + \"(IH)\"\n",
    "        \n",
    "    if (reason & QS):\n",
    "        read[2] = read[2] + \" \" + \"(QS)\"\n",
    "        \n",
    "    with open (outputFile, \"a\") as out:\n",
    "        out.write(read[0])\n",
    "        out.write(\"\\n\")\n",
    "        out.write(read[1])\n",
    "        out.write(\"\\n\")\n",
    "        out.write(read[2])\n",
    "        out.write(\"\\n\")\n",
    "        out.write(read[3])\n",
    "        out.write(\"\\n\")\n",
    "        \n",
    "def writeToGoodFile(outputFile, read):\n",
    "    with open (outputFile, \"a\") as out:\n",
    "        out.write(read[0])\n",
    "        out.write(\"\\n\")\n",
    "        out.write(read[1])\n",
    "        out.write(\"\\n\")\n",
    "        out.write(read[2])\n",
    "        out.write(\"\\n\")\n",
    "        out.write(read[3])\n",
    "        out.write(\"\\n\")\n",
    "        \n",
    "def convertPhred(letter):\n",
    "    return ord(letter)-33\n",
    "\n",
    "def meanQS(read):\n",
    "    total = 0\n",
    "    bp = 0\n",
    "    for base in read[3]:\n",
    "        total = total + convertPhred(base)\n",
    "        bp = bp + 1\n",
    "    return total/bp\n",
    "\n",
    "def compareMeans(read):\n",
    "    cutoffQS = 32\n",
    "    return meanQS(read) >= cutoffQS\n",
    "\n",
    "def containsN(index):\n",
    "    return \"N\" in index\n",
    "\n",
    "def readInLine(file):\n",
    "    line = file.readline()\n",
    "    return line.strip()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#10 Read subsets for each file\n",
    "#zcat /projects/bgmp/shared/2017_sequencing/1294_S1_L008_R1_001.fastq.gz | head -n40 > R1_10r.fq\n",
    "file10 = \"testFiles/R1_10r.fq\"\n",
    "file20 = \"testFiles/R2_10r.fq\"\n",
    "file30 = \"testFiles/R3_10r.fq\"\n",
    "file40 = \"testFiles/R4_10r.fq\"\n",
    "\n",
    "\n",
    "#100 Read subsets for each file\n",
    "#zcat /projects/bgmp/shared/2017_sequencing/1294_S1_L008_R1_001.fastq.gz | head -n400 > R1_100r.fq\n",
    "file100 = \"testFiles/R1_100r.fq\"\n",
    "file200 = \"testFiles/R2_100r.fq\"\n",
    "file300 = \"testFiles/R3_100r.fq\"\n",
    "file400 = \"testFiles/R4_100r.fq\"\n",
    "\n",
    "#1000 Read subsets for each file from the middle of the file\n",
    "#zcat /projects/bgmp/shared/2017_sequencing/1294_S1_L008_R4_001.fastq.gz | head -n4000000 | tail -n4000 > R4_1000r.fq\n",
    "file1000 = \"testFiles/R1_1000r.fq\"\n",
    "file2000 = \"testFiles/R2_1000r.fq\"\n",
    "file3000 = \"testFiles/R3_1000r.fq\"\n",
    "file4000 = \"testFiles/R4_1000r.fq\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reads:  1000\n",
      "N's  :  4\n",
      "IH   :  85\n",
      "QS   :  0\n",
      "\n",
      "\n",
      "% of reads for each index\n",
      "{'GTAGCGTA': 1.8, 'CGATCGAT': 1.6, 'GATCAAGG': 2.6, 'AACAGCGA': 2.6, 'TAGCCATG': 2.3, 'CGGTAATC': 1.3, 'CTCTGGAT': 8.6, 'TACCGGAT': 21.7, 'CTAGCTCA': 5.1, 'CACTTCAC': 1.3, 'GCTACTCT': 1.7, 'ACGATCAG': 2.1, 'TATGGCAC': 3.6, 'TGTTCCGT': 4.6, 'GTCCTAAG': 2.6, 'TCGACAAG': 1.2, 'TCTTCGAC': 13.2, 'ATCATGCG': 3.0, 'ATCGTGGT': 1.5, 'TCGAGAGT': 2.9, 'TCGGATTC': 0.8, 'GATCTTGC': 0.7, 'AGAGTCCA': 2.7, 'AGGATAGC': 1.8, 'SNP': 0.1}\n"
     ]
    }
   ],
   "source": [
    "fastqLines = 4\n",
    "end = False\n",
    "N = 0\n",
    "IH = 0\n",
    "QS = 0\n",
    "reads = 0\n",
    "\n",
    "fh1 = open(file1000, \"r\")\n",
    "fh2 = open(file2000, \"r\")\n",
    "fh3 = open(file3000, \"r\")\n",
    "fh4 = open(file4000, \"r\")\n",
    "\n",
    "read1 = [None] * 4\n",
    "indexR1 = [None] * 4\n",
    "indexR2 = [None] * 4\n",
    "read2 = [None] * 4\n",
    "\n",
    "indices = {\n",
    "    'GTAGCGTA': 0, 'CGATCGAT': 0, 'GATCAAGG': 0, 'AACAGCGA': 0,\n",
    "    'TAGCCATG': 0, 'CGGTAATC': 0, 'CTCTGGAT': 0, 'TACCGGAT': 0, \n",
    "    'CTAGCTCA': 0, 'CACTTCAC': 0, 'GCTACTCT': 0, 'ACGATCAG': 0, \n",
    "    'TATGGCAC': 0, 'TGTTCCGT': 0, 'GTCCTAAG': 0, 'TCGACAAG': 0,\n",
    "    'TCTTCGAC': 0, 'ATCATGCG': 0, 'ATCGTGGT': 0, 'TCGAGAGT': 0,\n",
    "    'TCGGATTC': 0, 'GATCTTGC': 0, 'AGAGTCCA': 0, 'AGGATAGC': 0,\n",
    "    'SNP': 0\n",
    "}\n",
    "\n",
    "while end != True:\n",
    "    flag = 0\n",
    "    for x in range(fastqLines):\n",
    "        read1[x] = readInLine(fh1)\n",
    "        if read1[x] == \"\":\n",
    "            end = True\n",
    "            break\n",
    "        indexR1[x] = readInLine(fh2)\n",
    "        indexR2[x] = readInLine(fh3)\n",
    "        read2[x] = readInLine(fh4)\n",
    "        \n",
    "        if x == 2:\n",
    "            read1[2] = read1[2] + \" \" + indexR1[1]\n",
    "            index1 = indexR1[1]\n",
    "            read2[2] = read2[2] + \" \" + indexR2[1]\n",
    "            index2 = indexR2[1]\n",
    "    if end == True:\n",
    "        break\n",
    "\n",
    "    reads = reads + 1 \n",
    "        \n",
    "    if (containsN(index1)) or (containsN(index2)):\n",
    "        flag = flag + 1\n",
    "        N = N + 1\n",
    "    \n",
    "    if not(matching(index1, index2)):\n",
    "        flag = flag + 2\n",
    "        IH = IH + 1\n",
    "    \n",
    "    if not(compareMeans(read1[1])) or not(compareMeans(read1[1])):\n",
    "        flag = flag + 4\n",
    "        QS = QS + 1\n",
    "    \n",
    "    if flag == 0:\n",
    "        if index1 in indices.keys():\n",
    "            indices[index1] = indices[index1] + 1\n",
    "            writeToGoodFile('outputFiles/%s_R1.fastq' % index1, read1)\n",
    "            writeToGoodFile('outputFiles/%s_R2.fastq' % index1, read2)\n",
    "        else:\n",
    "            flag = flag + 8\n",
    "            indices[\"SNP\"] = indices[\"SNP\"] + 1\n",
    "            writeToBadFile('outputFiles/rejects_R1.fastq', read1, flag)\n",
    "            writeToBadFile('outputFiles/rejects_R2.fastq', read1, flag)\n",
    "    else:\n",
    "        writeToBadFile('outputFiles/rejects_R1.fastq', read1, flag)\n",
    "        writeToBadFile('outputFiles/rejects_R2.fastq', read2, flag)\n",
    "\n",
    "for key in indices:\n",
    "    indices[key] = round(((indices[key] / reads) * 100), 2)\n",
    "        \n",
    "print(\"Reads: \", reads)\n",
    "print(\"N's  : \", N)\n",
    "print(\"IH   : \", IH)\n",
    "print(\"QS   : \", QS)\n",
    "print(\"\\n\")\n",
    "print(\"% of reads for each index\")\n",
    "print(indices)\n",
    "\n",
    "fh1.close()\n",
    "fh2.close()\n",
    "fh3.close()\n",
    "fh4.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "with open(\"outputFiles/summary.txt\", \"w\") as out:\n",
    "    out.write(\"Reads: \")\n",
    "    out.write(str(reads))\n",
    "    out.write(\"\\n\")\n",
    "    out.write(\"N's  : \")\n",
    "    out.write(str(N))\n",
    "    out.write(\"\\n\")\n",
    "    out.write(\"IH   : \")\n",
    "    out.write(str(IH))\n",
    "    out.write(\"\\n\")\n",
    "    out.write(\"QS   : \")\n",
    "    out.write(str(QS))\n",
    "    out.write(\"\\n\")\n",
    "    out.write(\"\\n\")\n",
    "    out.write(\"% of reads for each index\")\n",
    "    out.write(\"\\n\")\n",
    "    out.write(str(indices))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "indices = {\n",
    "    'GTAGCGTA': 0, 'CGATCGAT': 0, 'GATCAAGG': 0, 'AACAGCGA': 0,\n",
    "    'TAGCCATG': 0, 'CGGTAATC': 0, 'CTCTGGAT': 0, 'TACCGGAT': 0, \n",
    "    'CTAGCTCA': 0, 'CACTTCAC': 0, 'GCTACTCT': 0, 'ACGATCAG': 0, \n",
    "    'TATGGCAC': 0, 'TGTTCCGT': 0, 'GTCCTAAG': 0, 'TCGACAAG': 0,\n",
    "    'TCTTCGAC': 0, 'ATCATGCG': 0, 'ATCGTGGT': 0, 'TCGAGAGT': 0,\n",
    "    'TCGGATTC': 0, 'GATCTTGC': 0, 'AGAGTCCA': 0, 'AGGATAGC': 0,\n",
    "    'SNP': 0\n",
    "}\n",
    "def openFile(file)\n",
    "    file\n",
    "\n",
    "for key in indices:\n",
    "    openFile('%s_R1' % key)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
